File: src/libs/ThirdParty.VBSuppliers/Properties/ExpediaRapid/ExpediaRapidAPI.vb
Imports System.Net.Http
Imports System.Text
Imports Intuitive.Net
Imports Newtonsoft.Json
Imports ThirdParty.Abstractions.Constants
Imports ThirdParty.Abstractions.Models.Property.Booking
Imports ThirdParty.VBSuppliers.ExpediaRapid.SerializableClasses

Namespace ExpediaRapid
    Public Class ExpediaRapidAPI
        Implements IExpediaRapidAPI

        Private ReadOnly _httpclient As HttpClient

        Public Sub New(httpClient As HttpClient)
            _httpclient = httpClient
        End Sub

        Public Function GetDeserializedResponse(Of TResponse As {IExpediaRapidResponse, New})(propertyDetails As PropertyDetails, request As WebRequests.Request) As TResponse Implements IExpediaRapidAPI.GetDeserializedResponse

            Dim responseString As String = GetResponse(propertyDetails, request)
            Dim response As New TResponse

            If response.IsValid(request.ResponseString,
                                CInt(request.HTTPWebResponse.StatusCode)) Then

                Return JsonConvert.DeserializeObject(Of TResponse)(request.ResponseString)
            End If

            Return Nothing
        End Function

        Public Function GetResponse(propertyDetails As PropertyDetails, request As WebRequests.Request) As String Implements IExpediaRapidAPI.GetResponse
            request.Send(_httpclient)

            propertyDetails.Logs.AddNew(ThirdParties.EXPEDIARAPID, $"{request.LogFileName} Request", request.RequestString)
            propertyDetails.Logs.AddNew(ThirdParties.EXPEDIARAPID, $"{request.LogFileName} Response", request.ResponseString)

            Return request.ResponseString
        End Function

    End Class

End Namespace

File: src/libs/ThirdParty.CSSuppliers/ExpediaRapid/ExpediaRapidAPI.cs
namespace ThirdParty.CSSuppliers.ExpediaRapid
{
    using System.Net.Http;
    using ThirdParty.Constants;
    using ThirdParty.Models.Property.Booking;
    using ThirdParty.CSSuppliers.ExpediaRapid.SerializableClasses;
    using Microsoft.Extensions.Logging;
    using System.Threading.Tasks;

    public class ExpediaRapidAPI : IExpediaRapidAPI
    {
        private readonly HttpClient _httpclient;
        private readonly ILogger<ExpediaRapidAPI> _logger;

        public ExpediaRapidAPI(HttpClient httpClient, ILogger<ExpediaRapidAPI> logger)
        {
            _httpclient = httpClient;
            _logger = logger;
        }

        public async Task<TResponse> GetDeserializedResponseAsync<TResponse>(PropertyDetails propertyDetails, Intuitive.Net.WebRequests.Request request) where TResponse : IExpediaRapidResponse<TResponse>, new()
        {
            string responseString = await GetResponseAsync(propertyDetails, request);
            var response = new TResponse();

            (bool valid, response) = response.GetValidResults(responseString, (int)request.ResponseStatusCode);

            if (!valid)
            {
                response = default;
            }

            return response;
        }

        public async Task<string> GetResponseAsync(PropertyDetails propertyDetails, Intuitive.Net.WebRequests.Request request)
        {
            await request.Send(_httpclient, _logger);

            propertyDetails.Logs.AddNew(ThirdParties.EXPEDIARAPID, $"{request.LogFileName} Request", request.RequestString);
            propertyDetails.Logs.AddNew(ThirdParties.EXPEDIARAPID, $"{request.LogFileName} Response", request.ResponseString);

            return request.ResponseString;
        }
    }
}